File: templates/wasm-node-typescript/src/node.ts
/**
 * Flow-Like WASM Node Template — Main Node (TypeScript)
 *
 * Edit getDefinition() to define your node's pins and run() to implement the logic.
 */

import {
	type Context,
	type ExecutionResult,
	NodeDefinition,
	PinDefinition,
	PinType,
} from "@flow-like/wasm-sdk-typescript";

export function getDefinition(): NodeDefinition {
	const nd = new NodeDefinition(
		"my_custom_node_ts",
		"My Custom Node",
		"A template WASM node built with TypeScript",
		"Custom/WASM",
	);

	nd.addPin(PinDefinition.inputExec("exec"));
	nd.addPin(
		PinDefinition.inputPin("input_text", PinType.STRING, { defaultValue: "" }),
	);
	nd.addPin(
		PinDefinition.inputPin("multiplier", PinType.I64, { defaultValue: 1 }),
	);

	nd.addPin(PinDefinition.outputExec("exec_out"));
	nd.addPin(PinDefinition.outputPin("output_text", PinType.STRING));
	nd.addPin(PinDefinition.outputPin("char_count", PinType.I64));

	nd.addPermission("streaming");

	return nd;
}

export function run(ctx: Context): ExecutionResult {
	const inputText = ctx.getString("input_text", "") ?? "";
	const multiplier = ctx.getI64("multiplier", 1) ?? 1;

	ctx.debug(`Processing: '${inputText}' x ${multiplier}`);

	const outputText = inputText.repeat(Math.max(multiplier, 0));
	const charCount = outputText.length;

	ctx.streamText(`Generated ${charCount} characters`);

	ctx.setOutput("output_text", outputText);
	ctx.setOutput("char_count", charCount);

	return ctx.success();
}

File: templates/wasm-node-python/src/node.py
"""
Flow-Like WASM Node Template — Main Node

This file is the starting point for your custom node.
Edit get_definition() to define your node's pins and run() to implement the logic.
"""

from sdk import (
    Context,
    ExecutionResult,
    NodeDefinition,
    PinDefinition,
    PinType,
)


def get_definition() -> NodeDefinition:
    """Define the node's interface: name, category, inputs, and outputs."""
    nd = NodeDefinition(
        name="my_custom_node_py",
        friendly_name="My Custom Node",
        description="A template WASM node that demonstrates basic functionality",
        category="Custom/WASM",
    )

    nd.add_pin(PinDefinition.input_exec("exec"))
    nd.add_pin(PinDefinition.input_pin("input_text", PinType.STRING, default=""))
    nd.add_pin(PinDefinition.input_pin("multiplier", PinType.I64, default=1))

    nd.add_pin(PinDefinition.output_exec("exec_out"))
    nd.add_pin(PinDefinition.output_pin("output_text", PinType.STRING))
    nd.add_pin(PinDefinition.output_pin("char_count", PinType.I64))

    nd.add_permission("streaming")

    return nd


def run(ctx: Context) -> ExecutionResult:
    """Execute the node logic."""
    input_text = ctx.get_string("input_text", "")
    multiplier = ctx.get_i64("multiplier", 1)

    ctx.debug(f"Processing: '{input_text}' x {multiplier}")

    output_text = input_text * max(multiplier, 0)
    char_count = len(output_text)

    ctx.stream_text(f"Generated {char_count} characters")

    ctx.set_output("output_text", output_text)
    ctx.set_output("char_count", char_count)

    return ctx.success()
